### PR TITLE
Add initial support for sorted writer

### PR DESCRIPTION
**Pull request by kewang1024**

### CODE REVIEW

File: velox/common/config/SpillConfig.h [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  int32_t maxSpillLevel;
+
+  // Percentage of input batches to be spilled for testing. 0 means no
+  // spilling for test.
+  int32_t testSpillPct;
+
+  // CompressionKind when spilling, CompressionKind_NONE means no compression.
+  common::CompressionKind compressionKind;
+};
+} // namespace facebook::velox::common
```
Comment (xiaoxmeng) [inline velox/common/config/SpillConfig.h:110]: End of line.

File: velox/common/config/SpillConfig.h [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        aggregationPartitionBits(_aggregationPartitionBits),
+        maxSpillLevel(_maxSpillLevel),
+        testSpillPct(_testSpillPct),
+        compressionKind(common::stringToCompressionKind(_compressionKind)) {}
+
+  /// Returns the hash join spilling level with given 'startBitOffset'.
+  ///
+  /// NOTE: we advance (or right shift) the partition bit offset when goes to
+  /// the next level of recursive spilling.
+  int32_t joinSpillLevel(uint8_t startBitOffset) const;
```
Comment (xiaoxmeng) [inline velox/common/config/SpillConfig.h:56]: Can you add SpillConfigTest.cpp to move the current tests on these APIs from SpillerTest.cpp? Thanks!

File: velox/common/config/SpillConfig.h [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+ */
+#pragma once
+#include <stdint.h>
+#include <string.h>
+
+#include <folly/executors/CPUThreadPoolExecutor.h>
+#include "velox/common/compression/Compression.h"
+
+namespace facebook::velox::common {
+// Specifies the config for spilling.
```
Comment (xiaoxmeng) [inline velox/common/config/SpillConfig.h:24]: I assume there is no changes on this structure? Thanks!

File: velox/connectors/Connector.h [hunk 4]
```diff
@@ -345,7 +346,9 @@ class Connector {
       RowTypePtr inputType,
       std::shared_ptr<ConnectorInsertTableHandle> connectorInsertTableHandle,
       ConnectorQueryCtx* connectorQueryCtx,
-      CommitStrategy commitStrategy) = 0;
+      CommitStrategy commitStrategy,
+      tsan_atomic<bool>* nonReclaimableSection,
```
Comment (xiaoxmeng) [inline velox/connectors/Connector.h:350]: I think each hive data sink needs its own flag so let's don't pass this flag from writer operator? thanks!

File: velox/connectors/fuzzer/FuzzerConnector.h [hunk 5]
```diff
@@ -123,7 +123,9 @@ class FuzzerConnector final : public Connector {
       std::shared_ptr<
           ConnectorInsertTableHandle> /*connectorInsertTableHandle*/,
       ConnectorQueryCtx* /*connectorQueryCtx*/,
-      CommitStrategy /*commitStrategy*/) override final {
+      CommitStrategy /*commitStrategy*/,
+      tsan_atomic<bool>* nonReclaimableSection,
+      const common::SpillConfig* spillConfig) override final {
```
Comment (xiaoxmeng) [inline velox/connectors/fuzzer/FuzzerConnector.h:128]: ditto: nonReclaimableSection

Comment (xiaoxmeng) [inline velox/connectors/fuzzer/FuzzerConnector.h:128]: Let's move spillConfig to ConnectorQueryCtx? Thanks!

File: velox/connectors/hive/HiveDataSink.h [hunk 6]
```diff
@@ -466,6 +468,9 @@ class HiveDataSink : public DataSink {
   const std::unique_ptr<core::PartitionFunction> bucketFunction_;
   std::shared_ptr<dwio::common::WriterFactory> writerFactory_;
 
+  tsan_atomic<bool>* const nonReclaimableSection_;
+  const common::SpillConfig* spillConfig_;
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.h:472]:
```
const common::SpillConfig* const spillConfig_;
```

File: velox/connectors/hive/HiveDataSink.cpp [hunk 7]
```diff
@@ -444,15 +450,41 @@ uint32_t HiveDataSink::appendWriter(const HiveWriterId& id) {
...
            .connectorProperties = connectorProperties_,
            .pool = connectorQueryCtx_->memoryPool(),
            .metricLogger = dwio::common::MetricsLog::voidLog(),
            .stats = ioStats_.back().get()}),
-      options));
+      options);
+  auto sortedProperty = insertTableHandle_->bucketProperty()->sortedBy();
+  if (sortedProperty.size() > 0) {
+    std::vector<column_index_t> sortColumnIndices;
+    std::vector<CompareFlags> sortCompareFlags;
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.cpp:465]: reserve on sortColumnIndices and sortCompareFlags? Thanks!

File: velox/dwio/common/Writer.h [hunk 8]
```diff
@@ -41,7 +41,7 @@ class Writer {
    * Appends 'data' to writer. Data might still be in memory and not
    * yet written to the file.
    */
-  virtual void write(const VectorPtr& data) = 0;
+  virtual void write(const RowVectorPtr& data) = 0;
```
Comment (xiaoxmeng) [inline velox/dwio/common/Writer.h:44]: Why we need this change?

File: velox/exec/Driver.h [hunk 9]
```diff
@@ -240,7 +240,7 @@ struct DriverCtx {
       const std::string& operatorType);
 
   /// Builds the spill config for the operator with specified 'operatorId'.
-  std::optional<Spiller::Config> makeSpillConfig(int32_t operatorId) const;
+  std::optional<common::SpillConfig> makeSpillConfig(int32_t operatorId) const;
```
Comment (xiaoxmeng) [inline velox/exec/Driver.h:243]: Do you need common:: namespace prefix? Thanks!

File: velox/exec/SortBuffer.cpp [hunk 10]
```diff
@@ -158,6 +158,12 @@ RowVectorPtr SortBuffer::getOutput() {
   return output_;
 }
 
+void SortBuffer::abort() {
```
Comment (xiaoxmeng) [inline velox/exec/SortBuffer.cpp:161]: Remove this?

File: velox/exec/tests/TableWriteTest.cpp [hunk 11]
```diff
@@ -1044,6 +1071,15 @@ class BucketedTableOnlyWriteTest
...
+        testParams.push_back(TestParam{
+            fileFormat,
+            TestMode::kBucketed,
+            CommitStrategy::kNoCommit,
+            HiveBucketProperty::Kind::kPrestoNative,
+            true,
+            multiDrivers,
+            CompressionKind_ZSTD}
+                                 .value);
       }
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:1083]: We need sort test utility here to verify.

File: velox/common/config/SpillConfig.h (hunk 3 above)
Comment (kewang1024) [reply velox/common/config/SpillConfig.h:24]: right, just moved to another namespace

File: velox/connectors/Connector.h [hunk 12]
```diff
@@ -257,6 +262,14 @@ class ConnectorQueryCtx {
     return config_;
   }
 
+  const common::SpillConfig* const getSpillConfig() const {
```
Comment (xiaoxmeng) [inline velox/connectors/Connector.h:265]:
```
const common::SpillConfig* getSpillConfig()
```

File: velox/exec/TableWriter.cpp [hunk 13]
```diff
@@ -55,7 +58,11 @@ TableWriter::TableWriter(
   const auto& connectorId = tableWriteNode->insertTableHandle()->connectorId();
   connector_ = connector::getConnector(connectorId);
   connectorQueryCtx_ = operatorCtx_->createConnectorQueryCtx(
-      connectorId, planNodeId(), connectorPool_);
+      connectorId,
+      planNodeId(),
+      connectorPool_,
+      spillConfig_.has_value() ? &(spillConfig_.value()) : nullptr,
+      &numSpillRuns_);
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:65]: I think we can't pass numSpillRuns_ from operator to sort writer as it needs concurrency control. Let SortingWriter object to pass its own numSpillRuns to SortingWriter as for now.

File: velox/exec/tests/TableWriteTest.cpp [hunk 14]
```diff
@@ -758,7 +758,7 @@ class TableWriteTest : public HiveConnectorTestBase {
 
   // Verifies the bucketed file data by checking if the bucket id of each read
   // row is the same as the one encoded in the corresponding bucketed file name.
-  void verifyBucketedFileData(const std::filesystem::path& filePath) {
+  void verifyBucketedSortedFileData(const std::filesystem::path& filePath) {
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:761]: Let's just call it verifyBucketedFileData as sorting is optional property to verify internally.

File: velox/exec/tests/TableWriteTest.cpp [hunk 15]
```diff
@@ -786,6 +786,18 @@ class TableWriteTest : public HiveConnectorTestBase {
     for (const auto bucketId : bucketIds) {
       ASSERT_EQ(expectedBucketId, bucketId);
     }
+
+    // Verifies the sorting behavior
+    if (testParam_.bucketSort()) {
+      auto sortedColumn1 = resultVector->childAt(4)->asFlatVector<double>();
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:792]: Let's store sorted column indexes as sortColumnIndices_ in TableWriteTest in Setup() as numPartitionKeyValues_ does.

File: velox/exec/tests/TableWriteTest.cpp [hunk 16]
```diff
@@ -786,6 +786,18 @@ class TableWriteTest : public HiveConnectorTestBase {
...
+      auto sortedColumn2 = resultVector->childAt(1)->asFlatVector<int32_t>();
+      for (int i = 0; i < resultVector->size() - 1; ++i) {
+        ASSERT_LE(sortedColumn1->valueAt(i), sortedColumn1->valueAt(i + 1));
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:795]:
```
for (int i = 0; i < resultVector->size() - 1; ++i) {
  for (int sortColumnIndex : sortColumnIndices_) {
     if (sortedColumn1->valueAt(i) < sortedColumn1->valueAt(i + 1)) {
         break;
     }
     ASSERT_EQ(sortedColumn1->valueAt(i), sortedColumn1->valueAt(i + 1));
  }
```

**Review by xiaoxmeng — APPROVED** [review]
@kewang1024 LGTM % minors.

File: velox/connectors/hive/HiveDataSink.cpp [hunk 17]
```diff
@@ -462,6 +468,40 @@ uint32_t HiveDataSink::appendWriter(const HiveWriterId& id) {
   return writerIndexMap_[id];
 }
 
+std::unique_ptr<facebook::velox::dwio::common::Writer>
+HiveDataSink::maybeCreateBucketSortWriter(
+    std::unique_ptr<facebook::velox::dwio::common::Writer> writer) {
+  if (insertTableHandle_->bucketProperty() != nullptr &&
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.cpp:474]: How about
```
if (insertTableHandle_->bucketProperty() == nullptr) {
  return writer;
}

      insertTableHandle_->bucketProperty()->sortedBy().empty()) {
  return writer;
}
const auto& sortedProperty = insertTableHandle_->bucketProperty()->sortedBy();
if (sortedProperty.empty()) {
   return writer;
}
...
```

File: velox/connectors/hive/HiveDataSink.cpp [hunk 18]
```diff
@@ -462,6 +468,40 @@ uint32_t HiveDataSink::appendWriter(const HiveWriterId& id) {
...
+    auto sortBuffer = std::make_unique<exec::SortBuffer>(
+        inputType_,
+        sortColumnIndices,
+        sortCompareFlags,
+        1000, // todo batch size
+        connectorQueryCtx_->memoryPool(),
+        &nonReclaimableSection_,
+        numSpillRuns_,
+        spillConfig_);
+    writer = std::make_unique<dwio::common::SortingWriter>(
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.cpp:499]:
```
return std::make_unique<dwio::common::SortingWriter>();
```

File: velox/connectors/hive/HiveDataSink.h [hunk 19]
```diff
@@ -466,6 +470,12 @@ class HiveDataSink : public DataSink {
   const std::unique_ptr<core::PartitionFunction> bucketFunction_;
   std::shared_ptr<dwio::common::WriterFactory> writerFactory_;
 
+  // todo: combine with numSpillRuns_ in Operator
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.h:473]:
```
s/todo/TODO/
```

File: velox/connectors/hive/HiveDataSink.cpp [hunk 20]
```diff
@@ -462,6 +467,42 @@ uint32_t HiveDataSink::appendWriter(const HiveWriterId& id) {
...
+HiveDataSink::maybeCreateBucketSortWriter(
+    std::unique_ptr<facebook::velox::dwio::common::Writer> writer) {
+  if (insertTableHandle_->bucketProperty() == nullptr) {
+    return writer;
+  }
+  const auto& sortedProperty = insertTableHandle_->bucketProperty()->sortedBy();
+  if (sortedProperty.empty()) {
+    return writer;
+  }
+  std::vector<column_index_t> sortColumnIndices;
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.cpp:480]: How about we initialize these sorting properties once in ctor? e.g. sortColumnIndices_ and sortCompareFlags_

File: velox/dwio/common/SortingWriter.h [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+
+  /// No action because we need to accumulate all data and sort before data can
+  /// be flushed
+  virtual void flush() override;
+
+  virtual void close() override;
+
+  virtual void abort() override;
+
+  std::unique_ptr<exec::SortBuffer> sortBuffer_;
```
Comment (xiaoxmeng) [inline velox/dwio/common/SortingWriter.h:41]:
```
const std::unique_ptr<exec::SortBuffer> sortBuffer_;
```

File: velox/vector/BaseVector.h [hunk 22]
```diff
@@ -267,9 +267,12 @@ class BaseVector {
     return compare(other, index, otherIndex, CompareFlags()).value();
   }
 
-  // Returns < 0 if 'this' at 'index' is less than 'other' at 'otherIndex', 0 if
-  // equal and > 0 otherwise. If flags.nullHandlingMode is not NoStop, the
-  // function may returns std::nullopt if null encountered.
+  // When CompareFlags is ASCENDING, returns < 0 if 'this' at 'index' is less
```
Comment (xiaoxmeng) [inline velox/vector/BaseVector.h:270]: /// for public comment in header file. Thanks!

File: velox/dwio/common/SortingWriter.h (hunk 21 above)
Comment (kewang1024) [reply velox/dwio/common/SortingWriter.h:41]: We need to call reset on sortBuffer_ in `abort` function, so can't set it to const

**Review by xiaoxmeng — APPROVED** [review]
@kewang1024 Thanks for iterating!

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 can you squash the commits into one? Thanks!

**Pull request closed without merging**